apps/crawl-website/app/crawl_website.py

```diff
@@ -0,0 +1,347 @@
+                    "start_url": args.start_url,
+                    "max_documents": args.max_documents,
+                    "id": id_,
+                }
+            }
+        },
+        {
+            "CreateIndex": {
+                "index_type": "entity",
+                "class": "Crawl",
+                "property_key": "id",
+            }
+        },
+    ])
+    assert db.last_query_ok(), db.last_response
+
+    return id_, start_time
+
+
+def update_crawl(db, crawl_id, start_time):
+    """Update the Crawl entity with end time, duration, and number of documents"""
+    end_time = datetime.now(timezone.utc).isoformat()
+    duration = (datetime.fromisoformat(end_time) -
+                datetime.fromisoformat(start_time)).total_seconds()
+    logging.info(f"Ending Crawler at {end_time}, {duration} seconds")
+
+    response, _ = db.query([
+        {
+            "FindEntity": {
+                "with_class": "Crawl",
+                "constraints": {
+                    "id": ["==", crawl_id],
+                },
+                "_ref": 1,
+            },
+        },
+        {
+            "FindEntity": {
+                "with_class": "CrawlDocument",
+                "is_connected_to": {"ref": 1},
+                "results": {"count": True},
+            },
+        }
+    ])
+    assert db.last_query_ok(), db.last_response
+
+    n_documents = response[1]["FindEntity"]["count"] if len(
+        response) == 2 else 0
+    logging.info(f"Found {n_documents} documents")
+
+    db.query([{
+        "FindEntity": {
+            "with_class": "Crawl",
+            "constraints": {
+                "id": ["==", crawl_id],
+            },
+            "_ref": 1,
+        }
+    }, {
+        "UpdateEntity": {
+            "ref": 1,
+            "properties": {
+                "end_time": {"_date": end_time},
+                "n_documents": n_documents,
+                "duration": duration,
+            }
+        }
+    }])
+    assert db.last_query_ok(), db.last_response
+
+
+def main(args):
+    # The following line produces duplicate logs in the output.
+    # Presumably scrapy is also setting up logging.
+    # logging.basicConfig(level=args.log_level.upper())
+
+    db = CommonLibrary.create_connector()
+
+    crawl_id, start_time = create_crawl(db, args)
+    logging.info(f"Starting Crawler with ID: {crawl_id}")
+    process = CrawlerProcess(settings={
+        "APERTUREDB_PIPELINE_ARGS": args,
+        "APERTUREDB_CRAWL_ID": crawl_id,
+        "DOWNLOADER_MIDDLEWARES": {
+            ContentTypeFilterMiddleware: 543,
+        },
+        "ITEM_PIPELINES": {
+            ApertureDBPipeline: 1000,
+        },
+        "LOG_LEVEL": args.log_level.upper(),
+        "ALLOWED_CONTENT_TYPES": args.content_types.split(";"),
+        "CLOSESPIDER_ITEMCOUNT": args.max_documents,
+        "CONCURRENT_REQUESTS": args.concurrent_requests,
+        "CONCURRENT_REQUESTS_PER_DOMAIN": args.concurrent_requests_per_domain,
```

Comment:
We should also expose a handy parmeter : [DOWNLOAD_DELAY](https://docs.scrapy.org/en/latest/topics/settings.html#download-delay)

Reply:
Done